### ISSUE TITLE
Object spread not checking for strictNullChecks

### ISSUE THREAD

**Issue by vintem** [Microsoft/TypeScript#14799]
**TypeScript Version:**  2.2.1

**Code**

```ts
type Movie = {
	title: string;
	yearReleased: number;
}

const m = { title: "The Matrix", yearReleased: 1999 };
const x: Movie = { ...m, title: undefined };
```

**Expected behavior:**
When strictNullChecks = true it should not allow for title = undefined
**Actual behavior:**
It does not raise an error and I can have an object with a required property undefined.
It does work if I try to set null to title, the compiler will raise an error

**Comment by mprobst** [comment]
Similar, but might be a different bug:

```ts
const x: {a: number}|null = null;
const y = {...x};
```

With `strictNullChecks`, this should complain about `x` being possibly null at the spread location (with `downlevelIteration` this also gives a `TypeError`).

**Comment by mhegazy** [comment]
The semantics of object spread allows for skipping `null` and `undefined` values, see https://github.com/tc39/proposal-object-rest-spread/blob/master/Spread.md

**Comment by mprobst** [comment]
@mhegazy sorry, incorrect example, and I found that the usage site I ran into actually used `!` (so, self inflicted pain...).

**Comment by sandersn** [comment]
Fix is up at #15938
